contracts/extensions/CoinMachine.sol

```diff
@@ -140,43 +161,72 @@ contract CoinMachine is ColonyExtension {
     emit ExtensionInitialised();
   }
 
+  function setWhitelist(address _whitelist) public onlyRoot notDeprecated {
+    whitelist = _whitelist;
+
+    emit WhitelistSet(_whitelist);
+  }
+
   /// @notice Purchase tokens from Coin Machine.
   /// @param _numTokens The number of tokens to purchase
   function buyTokens(uint256 _numTokens) public payable notDeprecated {
     updatePeriod();
 
-    uint256 numTokens = min(min(_numTokens, maxPerPeriod - activeSold), tokensToSell);
+    require(
+      whitelist == address(0x0) || Whitelist(whitelist).isApproved(msg.sender),
+      "coin-machine-unauthorised"
+    );
+
+    uint256 tokenBalance = getTokenBalance();
+    uint256 numTokens = min(min(_numTokens, maxPerPeriod - activeSold), tokenBalance);
     uint256 totalCost = wmul(numTokens, activePrice);
 
     activeIntake = add(activeIntake, totalCost);
     activeSold = add(activeSold, numTokens);
-    tokensToSell = sub(tokensToSell, numTokens);
 
     assert(activeSold <= maxPerPeriod);
 
+    // Check if we've sold out
+    if (numTokens >= tokenBalance) {
+      soldOut = true;
+    }
+
     if (purchaseToken == address(0x0)) {
       require(msg.value >= totalCost, "coin-machine-insufficient-funds");
-      if (msg.value > totalCost) { msg.sender.transfer(msg.value - totalCost); }
+      if (msg.value > totalCost) { msg.sender.transfer(msg.value - totalCost); } // Refund any balance
+      payable(address(colony)).transfer(totalCost);
     } else {
-      require(ERC20Extended(purchaseToken).transferFrom(msg.sender, address(this), totalCost), "coin-machine-transfer-failed");
+      require(ERC20(purchaseToken).transferFrom(msg.sender, address(colony), totalCost), "coin-machine-transfer-failed");
     }
 
-    colony.mintTokensFor(msg.sender, numTokens);
+    require(ERC20(token).transfer(msg.sender, numTokens), "coin-machine-transfer-failed");
```

Comment:
Can we make these two require messages different so we're able to distinguish which failed in the event one does?